add abiility to dump openshift docker build strategy logs even if bui…

gabemontero opened the pull request:
…ld succeeds

/assign @stevekuznetsov 

per our quick confab on slack Steve

was able to test locally after logging into api.ci and creating a test project 

Admittedly, I took the path of least resistance wrt having to add the flag to those different ci-operator build types.  Did not see a ready made path to store the config in one place and have it leveraged by everybody.

stevekuznetsov commented:
/assign @smarterclayton @petr-muller @bbguimaraes @droslean

gabemontero commented:
Sure ... and this most likely is another episode of "jenkins the snowflake".

So, we produce 5 images out of https://github.com/openshift/jenkins.

The output of one of those images builds, the core jenkins image, is useful when the build completes successfully, for the ART/OSBS processing we have to do after the fact.

So I want to dump the build log for that docker strategy build regardless.  But I don't need it for the other 4 images openshift/jenkins produces.

Thus, I wanted to set a bit on this stanza:  https://github.com/openshift/release/blob/master/ci-operator/config/openshift/jenkins/openshift-jenkins-master.yaml#L56-L63

so the openshfit build logs are dumped.

I do NOT want to ever unconfigure.  I always want the build logs dumped when the succeed so I can gather the data I need for those non-api.ci activities I need to do.

stevekuznetsov commented:
OK, then that seems reasonable to me as long as the other maintainers don't think there is a preferred way for you to get at that data

stevekuznetsov commented:
Out of curiosity, what is it that you need the log for that you can't get by inspecting the output image?

gabemontero commented:
Our build includes downloads of the 100 or so jenkins plugins we have to install.  It dumps out the versions of each those plugins.

I need those for ART/OSBS later on.

Adding image labels for each of those plugins is too much metadata for the image ... makes the image inspect unusable (IMO and a few others as well).

And after it is built, uncracking all the plugin binaries within the image is a pain.

stevekuznetsov commented:
> And after it is built, uncracking all the plugin binaries within the image is a pain.

This is a shame. If you think scripting it is harder than adding this flag to ci-op, let's go with ci-op.

gabemontero commented:
conflict rebase pushed

stevekuznetsov commented:
Let's get a review from someone else, this seems fine to me

droslean commented:
It looks fine, but I would like to see the documentation updated in this PR as well. Furthermore, why we need to just print the logs in the build instead of including the build logs in the artifacts?

Personally, I prefer the build logs to be gathered in artifacts.
@smarterclayton thoughts?

gabemontero commented:
I would be fine with the successfl build logs being available in the artifacts as well.  That path just was not apparent to me.

Please provide pointers on how to do that if that is the path you decide upon.

I'll update the doc in a bit.

petr-muller commented on pkg/api/types.go:
```diff
@@ -505,6 +525,11 @@ type ImageSourcePath struct {
+	// AlwaysDumpBuildLog let's you choose to dump the OpenShift Docker
+	// build log regardless of success or failure.  The default is to
+	// only dump the build log on a failure.
+	AlwaysDumpBuildLog bool `json:"always_dump_build_log"`
```
@stevekuznetsov, should all these be `omitempty` so that we do not need to write all those falses next time we need to serialize ciop configs?

stevekuznetsov replied:
Good clal

petr-muller commented:
> I would be fine with the successfl build logs being available in the artifacts as well. That path just was not apparent to me.
> 
> Please provide pointers on how to do that if that is the path you decide upon.

It should be just a matter of saving files to `artifactDir` - you could create a directory there and put the build logs into it. It seems that currently only template (https://github.com/openshift/ci-operator/blob/master/pkg/steps/template.go#L37) and pod (https://github.com/openshift/ci-operator/blob/master/pkg/steps/pod.go#L40) have `artifactDir` member and source step doesn't (https://github.com/openshift/ci-operator/blob/master/pkg/steps/source.go#L59), so that step struct would need to get that member. Steps are created and populated here: https://github.com/openshift/ci-operator/blob/master/pkg/defaults/defaults.go#L121. Then I think it should be just a matter of writing files there.

It's probably slightly more work (but not much I think), but I agree w/ @droslean it's a little bit cleaner. I would /lgtm even the current version if you don't think it's worth the effort.

gabemontero commented:
I'm attempting to go down the artifacts path using @petr-muller 's pointers as a start

My first attempt with the updated binary running

`./ci-operator --config ../release/ci-operator/config/openshift/jenkins/openshift-jenkins-master.yaml --git-ref=openshift/jenkins@master --artifact-dir=./logs` 

copied a bunch of stuff into `./logs`, but not the build logs

Investigating

stevekuznetsov commented:
@gabemontero ping?

gabemontero commented:
On Tue, Mar 5, 2019 at 8:16 PM Steve Kuznetsov <notifications@github.com>
wrote:

> @gabemontero <https://github.com/gabemontero> ping?
>
yep got diverted by various interrupts

Hope to circle back to this tomorrow

Still trying to get the archive wiring such that it actually collects the
pod logs

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/ci-operator/pull/279#issuecomment-469926029>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADbadLecGxATUmljRBs0ddA0_euYKpIsks5vTxb6gaJpZM4bTC6G>
> .
>

gabemontero commented:
Looking more like I'll be delayed one more day

stevekuznetsov commented:
@gabemontero no biggie -- just wanted to poke the zombie PR

gabemontero commented:
I circled back to this today, but https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/openshift_jenkins/811/pull-ci-openshift-jenkins-master-images/93?log#log has prevented progress 

Updating jenkins, will try again

gabemontero commented:
will circle back after https://github.com/openshift/jenkins/pull/812 gets through

gabemontero commented:
quick fyi ... the stars have aligned, no blockers, and I'm getting back to this one

gabemontero commented:
Also, circling back to @petr-muller 's notes in https://github.com/openshift/ci-operator/pull/279#issuecomment-468353834 

turns out the pod/template uses of artifacts do not line up with openshift builds in the sense that there are not hook points to get at the build pod definition before it is launched, but you are able to do that with ci-operator's interactions with templates and generic pods.

As such, I'm looking at creating a new flavor to the artifact logic (perhaps a new addArtifact* method that takes in a buildClient) to account for these differences.

gabemontero commented:
OK ... got a pretty simple solution that is working

Verified via:
```
./ci-operator --config ../release/ci-operator/config/openshift/jenkins/openshift-jenkins-master.yaml  --artifact-dir /home/gmontero/tmpsave --namespace gabemontero-archive-test --git-ref openshift/jenkins@master
```

will push commit up in a bit

gabemontero commented:
ok `make check` does not seem to like `make format` results? ... investigating ..

gabemontero commented:
`make test` and `make check` and `make format` run clean for me ... I'm a bit confused 

but in the meantime ... @petr-muller  @stevekuznetsov @droslean ptal at the artifacts based solution

gabemontero commented:
will attempt to manually revert `make format` changes which seem to be making ci/prow/lint unhappy

stevekuznetsov commented:
What version of `go` are you on?

gabemontero commented:
```
go version go1.10.5 linux/amd64
```

gabemontero commented on pkg/steps/artifacts.go:
```diff
@@ -693,6 +694,35 @@ func gatherContainerLogsOutput(podClient PodClient, artifactDir, namespace, podN
 	return kerrors.NewAggregate(validationErrors)
 }
 
+// for gathering successful build logs to the artifacts, there is no way to augment the build spec
```
this should have said `augment the pod spec` ... will push update shortly

gabemontero replied:
correction to comment pushed

stevekuznetsov commented:
We're on 1.11 :)

gabemontero commented:
ah ok ... mystery solved - cool

let's circle back then to the change to "leverage archives" 

@petr-muller @droslean @stevekuznetsov ptal when you get the cycles

droslean commented on pkg/steps/source.go:
```diff
@@ -254,7 +255,15 @@ func handleBuild(buildClient BuildClient, build *buildapi.Build, dry bool) error
-	return waitForBuild(buildClient, build.Namespace, build.Name)
+	err := waitForBuild(buildClient, build.Namespace, build.Name)
+	if err == nil && len(artifactDir) > 0 {
+		aerr := gatherSuccessfulBuildLog(buildClient, artifactDir, build.Namespace, build.Name)
```
```suggestion
		aerr := gatherSuccessfulBuildLog(buildClient, artifactDir, build.Namespace, build.Name)
		if err:= gatherSuccessfulBuildLog(buildClient, artifactDir, build.Namespace, build.Name); err != nil{
		log.Printf("problem gathering successful build %s logs into artifacts: %v", build.Name, aerr)
		}
```

gabemontero commented on pkg/steps/source.go:
```diff
@@ -254,7 +255,18 @@ func handleBuild(buildClient BuildClient, build *buildapi.Build, dry bool) error
+		// log error but do not fail successful build
+		log.Printf("problem gathering successful build %s logs into artifacts: %v", build.Name, err)
+		}
+		if aerr != nil {
```
oop ... accepting @droslean 's commit did not delete the `aerr` block ... will address

gabemontero commented:
comment addressed @droslean  ... ptal

gabemontero commented on pkg/steps/source.go:
```diff
@@ -170,11 +171,11 @@ func buildFromSource(jobSpec *api.JobSpec, fromTag, toTag api.PipelineImageStrea
+						From:           from,
+						ForcePull:      true,
+						NoCache:        true,
+						Env:            []coreapi.EnvVar{},
```
dammit ....

gabemontero commented:
ok ... comments addressed and golang 1.10 vs,. 1.11 gofmt incompatibility adrressed .... @droslean ptal

droslean commented on pkg/steps/project_image.go:
```diff
@@ -161,13 +163,15 @@ func (s *projectDirectoryImageBuildStep) Description() string {
 		buildClient: buildClient,
 		imageClient: imageClient,
 		istClient:   istClient,
+		podClient:   podClient,
```
Where this is being used?

gabemontero replied:
forgot to clean this up when I had to pivot on how we integrate with the existing artifact.go structures ... will update momentarily - thanks

gabemontero replied:
update pushed @droslean

droslean commented on pkg/steps/artifacts.go:
```diff
@@ -693,6 +694,35 @@ func gatherContainerLogsOutput(podClient PodClient, artifactDir, namespace, podN
+func gatherSuccessfulBuildLog(buildClient BuildClient, artifactDir, namespace, buildName string) error {
+	// adding a subdir to the artifactDir path similar to downloadArtifacts adding the container-logs subdir
+	dir := filepath.Join(artifactDir, "openshift-docker-build-strategy-logs")
+	if err := os.MkdirAll(dir, 0750); err != nil {
```
Since other steps will call that function too, there is a chance that this directory we already exist, right?

gabemontero replied:
ah ... sure, if multiple steps are possible, I should filter out already exist errors ... will update shortly

gabemontero replied:
check that ... the godoc/code from os.MkDirAll ... it is benign and just returns if the dir already exists:

```
// If path is already a directory, MkdirAll does nothing
// and returns nil.
func MkdirAll(path string, perm FileMode) error {
	// Fast path: if we can tell whether path is a directory or file, stop with success or error.
	dir, err := Stat(path)
	if err == nil {
		if dir.IsDir() {
			return nil
		}
```

stevekuznetsov replied:
Yea `MkdirAll` === `mkdir -p`

gabemontero commented:
bump @stevekuznetsov @droslean @petr-muller 

can we merge this one?  or do we need more review/acceptance ?

stevekuznetsov commented:
This LGTM but I was hoping @petr-muller @droslean would take a final look and merge

droslean commented:
/lgtm
/hold
@petr-muller Do you want to take a final look?

petr-muller commented:
> @petr-muller Do you want to take a final look?

Will have a look today.

gabemontero commented:
thanks @stevekuznetsov @droslean @petr-muller !

Gonna create a test PR and see what we see in a couple

gabemontero commented:
Well the jenkins image build is hitting api.ci mirror pain again, but I hunted some other PRs and if you look at https://gcsweb-ci.svc.ci.openshift.org/gcs/origin-ci-test/pr-logs/pull/openshift_machine-api-operator/251/pull-ci-openshift-machine-api-operator-master-images/606/artifacts/openshift-docker-build-strategy-logs/  we look good.

droslean commented:
> Well the jenkins image build is hitting api.ci mirror pain again, but I hunted some other PRs and if you look at https://gcsweb-ci.svc.ci.openshift.org/gcs/origin-ci-test/pr-logs/pull/openshift_machine-api-operator/251/pull-ci-openshift-machine-api-operator-master-images/606/artifacts/openshift-docker-build-strategy-logs/ we look good.

Unfortunately, we are not good :(. The `src.log.gz` contains both of the logs

gabemontero commented:
Sure but I'm not sure I'm seeing that.

src log:

```
ulling image "registry.svc.ci.openshift.org/ci/clonerefs@sha256:fe003d8562593e9082dda696354c2a4dc2449b7784499c7c003dea67e0957ed2" ...
Replaced Dockerfile FROM image pipeline:root

Pulling image docker-registry.default.svc:5000/ci-op-5gf7c5m5/pipeline@sha256:99670ba0ed8240b7ba652717cdd23a406e48283dbdd0508388d1dfa9c372daea ...
--> FROM docker-registry.default.svc:5000/ci-op-5gf7c5m5/pipeline@sha256:99670ba0ed8240b7ba652717cdd23a406e48283dbdd0508388d1dfa9c372daea as 0
--> ENV "CLONEREFS_OPTIONS"="{\"git_user_email\":\"ci-robot@openshift.io\",\"git_user_name\":\"ci-robot\",\"log\":\"/dev/null\",\"refs\":[{\"org\":\"openshift\",\"repo\":\"machine-api-operator\",\"base_ref\":\"master\",\"base_sha\":\"9d4bb9c553e600e5b2c4a7c2a681c17080268e78\",\"pulls\":[{\"number\":251,\"author\":\"ingvagabund\",\"sha\":\"5396196ffeb43d2c6b15907116dc763b7f2062d5\"}],\"path_alias\":\"github.com/openshift/machine-api-operator\"}],\"src_root\":\"/go\"}"
--> ADD ./app.binary /clonerefs
--> RUN umask 0002 && /clonerefs && find /go/src -type d -not -perm -0775 | xargs -r chmod g+xw
{"component":"clonerefs","level":"info","msg":"Cloning refs","refs":{"org":"openshift","repo":"machine-api-operator","base_ref":"master","base_sha":"9d4bb9c553e600e5b2c4a7c2a681c17080268e78","pulls":[{"number":251,"author":"ingvagabund","sha":"5396196ffeb43d2c6b15907116dc763b7f2062d5"}],"path_alias":"github.com/openshift/machine-api-operator"},"time":"2019-03-26T13:51:42Z"}
{"command":"mkdir -p /go/src/github.com/openshift/machine-api-operator","component":"clonerefs","error":null,"level":"info","msg":"Ran command","output":"","time":"2019-03-26T13:51:42Z"}
{"command":"git init","component":"clonerefs","error":null,"level":"info","msg":"Ran command","output":"Initialized empty Git repository in /go/src/github.com/openshift/machine-api-operator/.git/\n","time":"2019-03-26T13:51:42Z"}
{"command":"git config user.name ci-robot","component":"clonerefs","error":null,"level":"info","msg":"Ran command","output":"","time":"2019-03-26T13:51:42Z"}
{"command":"git config user.email ci-robot@openshift.io","component":"clonerefs","error":null,"level":"info","msg":"Ran command","output":"","time":"2019-03-26T13:51:42Z"}
{"command":"git fetch https://github.com/openshift/machine-api-operator.git --tags --prune","component":"clonerefs","error":null,"level":"info","msg":"Ran command","output":"From https://github.com/openshift/machine-api-operator\n * [new tag]         v0.1.0     -\u003e v0.1.0\n * [new tag]         v0.2.0     -\u003e v0.2.0\n","time":"2019-03-26T13:51:45Z"}
{"command":"git fetch https://github.com/openshift/machine-api-operator.git master","component":"clonerefs","error":null,"level":"info","msg":"Ran command","output":"From https://github.com/openshift/machine-api-operator\n * branch            master     -\u003e FETCH_HEAD\n","time":"2019-03-26T13:51:47Z"}
{"command":"git checkout 9d4bb9c553e600e5b2c4a7c2a681c17080268e78","component":"clonerefs","error":null,"level":"info","msg":"Ran command","output":"Note: checking out '9d4bb9c553e600e5b2c4a7c2a681c17080268e78'.\n\nYou are in 'detached HEAD' state. You can look around, make experimental\nchanges and commit them, and you can discard any commits you make in this\nstate without impacting any branches by performing another checkout.\n\nIf you want to create a new branch to retain commits you create, you may\ndo so (now or later) by using -b with the checkout command again. Example:\n\n  git checkout -b new_branch_name\n\nHEAD is now at 9d4bb9c... Merge pull request #235 from russellb/baremetal\n","time":"2019-03-26T13:51:49Z"}
{"command":"git branch --force master 9d4bb9c553e600e5b2c4a7c2a681c17080268e78","component":"clonerefs","error":null,"level":"info","msg":"Ran command","output":"","time":"2019-03-26T13:51:49Z"}
{"command":"git checkout master","component":"clonerefs","error":null,"level":"info","msg":"Ran command","output":"Switched to branch 'master'\n","time":"2019-03-26T13:51:49Z"}
{"component":"clonerefs","error":"strconv.Atoi: parsing \"1553606064\\n\": invalid syntax","level":"error","msg":"Failed to parse timestamp \"1553606064\\n\"","time":"2019-03-26T13:51:49Z"}
{"command":"git fetch https://github.com/openshift/machine-api-operator.git pull/251/head","component":"clonerefs","error":null,"level":"info","msg":"Ran command","output":"From https://github.com/openshift/machine-api-operator\n * branch            refs/pull/251/head -\u003e FETCH_HEAD\n","time":"2019-03-26T13:51:50Z"}
{"command":"git merge --no-ff 5396196ffeb43d2c6b15907116dc763b7f2062d5","component":"clonerefs","error":null,"level":"info","msg":"Ran command","output":"Auto-merging README.md\nMerge made by the 'recursive' strategy.\n README.md              | 10 +++++-----\n config/machineset.yaml | 53 ++++++++++++++++++++++++++++++++++++++++++++++++++\n 2 files changed, 58 insertions(+), 5 deletions(-)\n create mode 100644 config/machineset.yaml\n","time":"2019-03-26T13:51:50Z"}
{"command":"git submodule update --init --recursive","component":"clonerefs","error":null,"level":"info","msg":"Ran command","output":"","time":"2019-03-26T13:51:50Z"}
{"component":"clonerefs","level":"info","msg":"Finished cloning refs","time":"2019-03-26T13:51:50Z"}
--> WORKDIR /go/src/github.com/openshift/machine-api-operator/
--> ENV GOPATH=/go
--> RUN git submodule update --init
--> ENV "OPENSHIFT_BUILD_NAME"="src" "OPENSHIFT_BUILD_NAMESPACE"="ci-op-5gf7c5m5"
--> LABEL "io.openshift.build.name"="src" "io.openshift.build.namespace"="ci-op-5gf7c5m5"
--> Committing changes to temp.builder.openshift.io/ci-op-5gf7c5m5/src:4629baeb ...
--> Done

Pushing image docker-registry.default.svc:5000/ci-op-5gf7c5m5/pipeline:src ...
Pushed 2/3 layers, 69% complete
Pushed 3/3 layers, 100% complete
Push successful
```

machine-api-operator log

```
Pulling image "docker-registry.default.svc:5000/ci-op-5gf7c5m5/pipeline@sha256:ddb40521ed1a4f55e9bb4a2d7c7f30819c3e6830bb39666c5408bb6d6f4a144f" ...
Replaced Dockerfile FROM image registry.svc.ci.openshift.org/openshift/origin-v4.0:base

Pulling image docker-registry.default.svc:5000/ci-op-5gf7c5m5/pipeline@sha256:3eea282ed886c5f958b44856a06bb01b90fb684190cf9f5da6eaf2f2ddd99ee8 ...

Pulling image registry.svc.ci.openshift.org/openshift/release:golang-1.10 ...
--> FROM registry.svc.ci.openshift.org/openshift/release:golang-1.10 as builder
--> WORKDIR /go/src/github.com/openshift/machine-api-operator
--> COPY . .
--> RUN NO_DOCKER=1 make build
./hack/go-build.sh machine-api-operator
Using version from git...
Building github.com/openshift/machine-api-operator/cmd/machine-api-operator (v0.1.0-268-g52799f23-dirty)
./hack/go-build.sh nodelink-controller
Using version from git...
Building github.com/openshift/machine-api-operator/cmd/nodelink-controller (v0.1.0-268-g52799f23-dirty)
./hack/go-build.sh machine-healthcheck
Using version from git...
Building github.com/openshift/machine-api-operator/cmd/machine-healthcheck (v0.1.0-268-g52799f23-dirty)
--> FROM docker-registry.default.svc:5000/ci-op-5gf7c5m5/pipeline@sha256:3eea282ed886c5f958b44856a06bb01b90fb684190cf9f5da6eaf2f2ddd99ee8 as 1
--> COPY --from=builder /go/src/github.com/openshift/machine-api-operator/owned-manifests owned-manifests
--> COPY --from=builder /go/src/github.com/openshift/machine-api-operator/install manifests
--> COPY --from=builder /go/src/github.com/openshift/machine-api-operator/bin/machine-api-operator .
--> COPY --from=builder /go/src/github.com/openshift/machine-api-operator/bin/nodelink-controller .
--> COPY --from=builder /go/src/github.com/openshift/machine-api-operator/bin/machine-healthcheck .
--> LABEL io.openshift.release.operator=true
--> ENV "OPENSHIFT_BUILD_NAME"="machine-api-operator" "OPENSHIFT_BUILD_NAMESPACE"="ci-op-5gf7c5m5"
--> LABEL "io.openshift.build.name"="machine-api-operator" "io.openshift.build.namespace"="ci-op-5gf7c5m5" "io.openshift.build.name"="" "io.openshift.build.commit.ref"="" "io.openshift.build.commit.message"="" "io.openshift.build.commit.author"="" "io.openshift.build.commit.date"="" "vcs-type"="" "vcs-url"="" "io.openshift.build.commit.id"="" "io.openshift.build.source-location"="" "io.openshift.build.source-context-dir"="" "vcs-ref"="" "io.openshift.build.namespace"=""
--> Committing changes to temp.builder.openshift.io/ci-op-5gf7c5m5/machine-api-operator:487cb51e ...
--> Done

Pushing image docker-registry.default.svc:5000/ci-op-5gf7c5m5/pipeline:machine-api-operator ...
Pushed 4/5 layers, 81% complete
Pushed 5/5 layers, 100% complete
Push successful
```

they look different to me

gabemontero commented:
And I only see 1 file in each .gz

droslean commented:
@gabemontero sorry, for some reason my `file-roller` was appending the temp files... ok so we are good :)

gabemontero commented:
On Tue, Mar 26, 2019 at 10:33 AM Nikolaos Moraitis <notifications@github.com>
wrote:

> @gabemontero <https://github.com/gabemontero> sorry, for some reason my
> file-roller was appending the temp files... ok so we are good :)
>
whew ok cool :-)

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/ci-operator/pull/279#issuecomment-476673865>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADbadApNBO-yObVo4gmxfj_DkgMgtn2mks5vai_TgaJpZM4bTC6G>
> .
>

smarterclayton commented on pkg/steps/artifacts.go:
```diff
@@ -693,6 +694,35 @@ func gatherContainerLogsOutput(podClient PodClient, artifactDir, namespace, podN
+// around download/copy from the artifacts container's volume mount and multiple pods are avoided.
+func gatherSuccessfulBuildLog(buildClient BuildClient, artifactDir, namespace, buildName string) error {
+	// adding a subdir to the artifactDir path similar to downloadArtifacts adding the container-logs subdir
+	dir := filepath.Join(artifactDir, "openshift-docker-build-strategy-logs")
```
I don't like this name.  I'd prefer we shorten this to `build-logs`.

gabemontero replied:
https://github.com/openshift/ci-operator/pull/303
